types/mempool/priority.go

```diff
@@ -0,0 +1,234 @@
+func NewPriorityMempool(opts ...PriorityMempoolOption) Mempool {
+	mp := &priorityMempool{
+		priorityIndex: huandu.New(huandu.LessThanFunc(txKeyLess)),
+		senderIndices: make(map[string]*huandu.SkipList),
+		scores:        make(map[txKey]int64),
+	}
+	for _, opt := range opts {
+		opt(mp)
+	}
+	return mp
+}
+
+// Insert attempts to insert a Tx into the app-side mempool in O(log n) time, returning an error if unsuccessful.
+// Sender and nonce are derived from the transaction's first signature.
+// Transactions are unique by sender and nonce.
+// Inserting a duplicate tx is an O(log n) no-op.
+// Inserting a duplicate tx with a different priority overwrites the existing tx, changing the total order of
+// the mempool.
+func (mp *priorityMempool) Insert(ctx sdk.Context, tx Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	tk := txKey{nonce: nonce, priority: ctx.Priority(), sender: sender}
+
+	senderIndex, ok := mp.senderIndices[sender]
+	if !ok {
+		senderIndex = huandu.New(huandu.LessThanFunc(func(a, b any) int {
+			return huandu.Uint64.Compare(b.(txKey).nonce, a.(txKey).nonce)
+		}))
+		// initialize sender index if not found
+		mp.senderIndices[sender] = senderIndex
+	}
+
+	// Since senderIndex is scored by nonce, a changed priority will overwrite the existing txKey.
+	senderIndex.Set(tk, tx)
+
+	// Since mp.priorityIndex is scored by priority, then sender, then nonce, a changed priority will create a new key,
+	// so we must remove the old key and re-insert it to avoid having the same tx with different priorityIndex indexed
+	// twice in the mempool.  This O(log n) remove operation is rare and only happens when a tx's priority changes.
+	sk := txKey{nonce: nonce, sender: sender}
+	if oldScore, txExists := mp.scores[sk]; txExists {
+		mp.priorityIndex.Remove(txKey{nonce: nonce, priority: oldScore, sender: sender})
+	}
+	mp.scores[sk] = ctx.Priority()
+	mp.priorityIndex.Set(tk, tx)
+
+	return nil
+}
+
+// Select returns a set of transactions from the mempool, ordered by priority and sender-nonce in O(n) time.
+// The passed in list of transactions are ignored.  This is a readonly operation, the mempool is not modified.
+// maxBytes is the maximum number of bytes of transactions to return.
+func (mp *priorityMempool) Select(_ [][]byte, maxBytes int64) ([]Tx, error) {
+	var selectedTxs []Tx
+	var txBytes int64
+	senderCursors := make(map[string]*huandu.Element)
+
+	// start with the highest priority sender
+	priorityNode := mp.priorityIndex.Front()
+	for priorityNode != nil {
+		priorityKey := priorityNode.Key().(txKey)
+		nextHighestPriority, nextPriorityNode := nextPriority(priorityNode)
+		sender := priorityKey.sender
+		senderTx := mp.fetchSenderCursor(senderCursors, sender)
```

Comment:
```suggestion
		senderNode := mp.fetchSenderCursor(senderCursors, sender)
```

Reply:
I actually don't think you need this function at all if you initially populate `senderCursors`

```go
senderCursors[sender] = mp.senderIndices[sender].Front()
```

Reply:
I don't think eager iteration of every sender sub-tree is a good idea.  What if we never reach them due to `maxBytes` being reached?